File: 5. Imported Model/MetalTryOut-Objc/Shaders/METLTexture.m
/*
 Copyright (C) 2014 Apple Inc. All Rights Reserved.
 See LICENSE.txt for this sample’s licensing information
 
 Abstract:
 
  Simple Utility class for creating a 2d texture
  
 */

#import "METLTexture.h"

@implementation METLTexture

- (instancetype) initWithResourceName:(NSString *)name
                                  ext:(NSString *)ext
{
    NSString *path = [[NSBundle mainBundle] pathForResource:name
                                                     ofType:ext];
    
    if(!path)
    {
        return nil;
    } // if
    
    self = [super init];
    
    if (self)
    {
        _path     = path;
        _width    = 0;
        _height   = 0;
        _depth    = 1;
        _format   = MTLPixelFormatRGBA8Unorm;
        _target   = MTLTextureType2D;
        _texture  = nil;
        _hasAlpha = NO;
    } // if
    
    return self;
} // initWithResourceName

- (BOOL) finalize:(id <MTLDevice>)device
{
    return [self finalize:device
                     flip:YES];
} // finalize

- (void) dealloc
{
    _path    = nil;
    _texture = nil;
} // dealloc

// assumes png file
- (BOOL) finalize:(id <MTLDevice>)device
             flip:(BOOL)flip
{
    UIImage *pImage = [UIImage imageWithContentsOfFile:_path];
    
    if(!pImage)
    {
        pImage = nil;
        
        return NO;
    } // if
    
    CGColorSpaceRef pColorSpace = CGColorSpaceCreateDeviceRGB();
    
    if(!pColorSpace)
    {
        pImage = nil;
        
        return NO;
    } // if
    
    self.width  = (uint32_t)CGImageGetWidth(pImage.CGImage);
    self.height = (uint32_t)CGImageGetHeight(pImage.CGImage);
    
    uint32_t width    = _width;
    uint32_t height   = _height;
    uint32_t rowBytes = width * 4;
    
    CGContextRef pContext = CGBitmapContextCreate(NULL,
                                                  width,
                                                  height,
                                                  8,
                                                  rowBytes,
                                                  pColorSpace,
                                                  (CGBitmapInfo)kCGImageAlphaPremultipliedLast);
    
    CGColorSpaceRelease(pColorSpace);
    
    if(!pContext)
    {
        return NO;
    } // if
    
    CGRect bounds = CGRectMake(0.0f, 0.0f, width, height);
    
    CGContextClearRect(pContext, bounds);
    
    // Vertical Reflect
    if(flip)
    {
        CGContextTranslateCTM(pContext, 0, height);
        CGContextScaleCTM(pContext, 1.0, -1.0);
    } // if
    
    CGContextDrawImage(pContext, bounds, pImage.CGImage );
    
    pImage = nil;
    
    MTLTextureDescriptor *pTexDesc = [MTLTextureDescriptor texture2DDescriptorWithPixelFormat:MTLPixelFormatRGBA8Unorm
                                                                                        width:width
                                                                                       height:height
                                                                                    mipmapped:YES];
    self.target  = pTexDesc.textureType;
    self.texture = [device newTextureWithDescriptor:pTexDesc];
    
    pTexDesc = nil;
    
    if(!self.texture)
    {
        CGContextRelease(pContext);
        
        return NO;
    } // if
    
    const void *pPixels = CGBitmapContextGetData(pContext);
    
    [self.texture replaceRegion:MTLTextureRegionMake2D(0, 0, width, height)
                    mipmapLevel:0
                      withBytes:pPixels
                    bytesPerRow:rowBytes];
    
    CGContextRelease(pContext);
    
    return YES;
} // finalize

@end

File: 5. Imported Model/MetalTryOut-Objc/Shaders/METLTexture.swift
//
//  METLTexture.swift
//  Textured Cube
//
//  Created by Andrew K. on 6/29/14.
//  Copyright (c) 2014 Andrew Kharchyshyn. All rights reserved.
//

import UIKit
import Metal

class METLTexture: NSObject {
   
    var texture: MTLTexture?
    var target: MTLTextureType
    var width: UInt32
    var height: UInt32
    var depth: UInt32
    var format: MTLPixelFormat
    var hasAlpha: Bool
    var path: String
    
    init(name: String, extention: String)
    {
        var qPath = NSBundle.mainBundle().pathForResource(name, ofType: extention);
        
        path     = qPath
        width    = 0
        height   = 0
        depth    = 1
        format   = MTLPixelFormat.FormatRGBA8Unorm
        target   = MTLTextureType.Type2D
        texture  = nil
        hasAlpha = false
        
        super.init()
    }
    
    func finalizeTexture(device:MTLDevice) -> Bool
    {
        return finalizeTexture(device, flip: true)
    }
    
    func finalizeTexture(device: MTLDevice, flip: Bool) -> Bool
    {
        var pImage: UIImage? = UIImage(contentsOfFile: path)
        if pImage == nil
        {
            return false
        }
        
        var pColorSpace:CGColorSpaceRef  = CGColorSpaceCreateDeviceRGB()
        
        if pColorSpace == nil
        {
            pImage = nil
            return false
        } // if
        
        self.width  =  UInt32(CGImageGetWidth(pImage?.CGImage))
        self.height =  UInt32(CGImageGetHeight(pImage?.CGImage))
        
        var width = self.width
        var height   = self.height
        var rowBytes = width * 4
        
        var info:CGBitmapInfo = CGBitmapInfo(CGImageAlphaInfo.PremultipliedLast.toRaw())
        
        var pContext:CGContextRef  = CGBitmapContextCreate(nil, UInt(width), UInt(height), 8, UInt(rowBytes), pColorSpace, info)
        
        

        CGColorSpaceRelease(pColorSpace)
        
        if pContext == nil
        {
            return false
        }
        
        var bounds = CGRectMake(0.0, 0.0, CGFloat(width), CGFloat(height));
        
        CGContextClearRect(pContext, bounds)
        
        // Vertical Reflect
        if(flip)
        {
            CGContextTranslateCTM(pContext, CGFloat(width), CGFloat(height));
            CGContextScaleCTM(pContext, -1.0, -1.0);
        } // if
        
        CGContextDrawImage(pContext, bounds, pImage!.CGImage);
        
        pImage = nil;
        
        var pTexDesc = MTLTextureDescriptor.texture2DDescriptorWithPixelFormat(MTLPixelFormat.FormatRGBA8Unorm, width: Int(width), height: Int(height), mipmapped: false)
        self.target = pTexDesc.textureType
        self.texture = device.newTextureWithDescriptor(pTexDesc)
        
        pTexDesc = nil;
        
        if self.texture == nil
        {
            CGContextRelease(pContext)
            return false
        }
        
        var pPixels: COpaquePointer = CGBitmapContextGetData(pContext);
        self.texture?.replaceRegion(MTLTextureRegionMake2D(0, 0, Int(width), Int(height)), mipmapLevel: 0, withBytes: pPixels, bytesPerRow: Int(rowBytes))
        
            
        CGContextRelease(pContext)
        
        return true
    }
    
}
